pinot-core/src/test/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManagerTest.java

```diff
@@ -609,6 +639,43 @@ public void testFileRemovedDuringOnlineTransition()
     segmentDataManager._state.set(segmentDataManager, LLRealtimeSegmentDataManager.State.HOLDING);
     segmentDataManager.goOnlineFromConsuming(metadata);
     Assert.assertFalse(new File(segTarFileName).exists());
+    segmentDataManager.destroy();
+  }
+
+  @Test
+  public void testOnlyOneSegmentHoldingTheSemaphoreForParticularPartition()
+      throws Exception {
+    long timeout = 2_000L;
+    FakeLLRealtimeSegmentDataManager firstSegmentDataManager = createFakeSegmentManager();
+    Assert.assertTrue(firstSegmentDataManager.getAcquireConsumerSemaphore().get());
+    Assert.assertEquals(firstSegmentDataManager.getPartitionConsumerSemaphore().availablePermits(), 0);
+
+    // Release semaphore after timeout.
+    Thread releaseSemaphore = new Thread(() -> {
+      try {
+        Thread.sleep(timeout);
```

Comment:
Can you use barriers and avoid the sleep ? Is this test prone to being flaky?

Reply:
No, the permit of the semaphore will be released in the finally block. This test won't be flaky.